---
title: "Epigenome-wide association study for vitamin K response to supplementation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=F, message=F, warning=F, cache.path="../cache2/")
library(tidyverse)
library(knitr)
library(car)
library(doParallel)
library(foreach)
library(itertools)
library(IlluminaHumanMethylationEPICanno.ilm10b2.hg19)
library(qqman)
library(lmerTest)
library(kableExtra)
```

```{r load-data}
betas <- readRDS("../int/betas.qc.norm.filt.rds")

metadata <- read_csv("../data/phen/finalTbl_withPIVKAandOtherVKMeasures.csv")
sampleSheet <- read_csv("../int/sampleSheet.csv")

load("../int/pca.RData")
cellCounts <- readRDS("../int/estCellCounts.rds")
load("../int/CPACOR.RData")
```

```{r clean-data}
mvals <- minfi::logit2(betas)

cellCounts <- as.data.frame(cellCounts) %>%
  rownames_to_column(var="sampleKey")

nmd <- metadata %>%  # nmd = "non-methylation data"
  mutate(responder=ifelse(K_responder_YN == "Yes", T, F)) %>%
  inner_join(sampleSheet, by="hnrcid") %>%
  mutate(sampleKey=paste(Sentrix_ID, Sentrix_Position, sep="_"),
         row=as.integer(substr(Sentrix_Position, 3, 3))) %>%
  inner_join(PCs, by="sampleKey") %>%
  inner_join(cellCounts, by="sampleKey") %>%
  inner_join(CP_PCs, by="sampleKey")
nmd <- nmd[match(colnames(betas), nmd$sampleKey), ]
stopifnot(all(nmd$sampleKey == colnames(betas)))

epic_anno <- data.frame(getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b2.hg19),
                        stringsAsFactors=F) %>%
  filter(!(chr %in% c("chrX", "chrY"))) %>%
  mutate(chr=as.integer(gsub("chr", "", chr)))
```

```{r helper-funcs}
make_qqplot <- function(p_vec, plotTitle="Title") {
  p_vec <- p_vec[!is.na(p_vec)]
  qqplot(-log10(1:length(p_vec) / length(p_vec)), -log10(p_vec), pch=".", 
         main=plotTitle, xlab="Expected (-logP)", ylab="Observed (-logP)")
  abline(0, 1, col="red")
}

make_large_qqplot <- function(p_vec, plotTitle="Title") {
  p_vec <- p_vec[!is.na(p_vec)]
  qqplot(-log10(1:length(p_vec) / length(p_vec)), -log10(p_vec), pch=".", 
         cex=4,
         main=plotTitle, xlab="Expected (-logP)", ylab="Observed (-logP)")
  abline(0, 1, col="red")
}

gControl <- function(p_vals) {
  # See van Iterson 2017 methods and/or Lehne 2015 code for details on genomic control for EWAS
  # Below is modeled after Lehne 2015
  lambda <- median(qchisq(p_vals, df=1, lower.tail=F), 
                            na.rm=T) / qchisq(0.5, df=1)
  round(lambda, 2)
}
```

```{r gwas-integration-prep}
if (!file.exists("../int/hassan_gwas_anno.rds")) {
  library(data.table)
  hassan_gwas <- fread("../data/literature/GWASVITK_M1_GC1.txt") %>%
    as_tibble() %>%
    select(MarkerName, Effect, `P-value`)
  
  snp_anno <- fread("../int/snp_annot_hg19_nodups.txt")
  snp_anno_df <- as_tibble(snp_anno) %>%
    setNames(c("chr", "pos", "SNP", "A1", "A2", "concat")) %>%
    select(SNP, chr, pos)
  
  hassan_gwas_anno <- hassan_gwas %>%
    inner_join(snp_anno_df, by=c("MarkerName"="SNP"))
  saveRDS(hassan_gwas_anno, "../int/hassan_gwas_anno.rds")
}

hassan_gwas_anno <- readRDS("../int/hassan_gwas_anno.rds")
hassan_gwas_anno <- mutate(hassan_gwas_anno, chr=as.integer(chr)) %>%
  mutate(negLogP=-log10(`P-value`)) %>%
  na.omit()

gwas_ranges <- GRanges(seqnames=hassan_gwas_anno$chr, 
                       ranges=IRanges(start=hassan_gwas_anno$pos, 
                                      end=hassan_gwas_anno$pos))
```

# Responder EWAS

```{r responder-ewas-prep}
run_responder_model <- function(probeData, covarData, formula_string) {
  CpG <- rownames(probeData)
  regData <- cbind(meth=as.vector(probeData), covarData)
  tryCatch({
    lm_fit <- lm(as.formula(formula_string), data=regData)
    coefs <- summary(lm_fit)$coef["responderTRUE", c("Estimate", "t value", "Pr(>|t|)")]
    setNames(c(CpG, coefs), c("CpG", "beta", "t", "p"))
  }, error=function(e) rep(NA, 3))
}

cpPC_string <- paste(paste0("cpPC", 1:10), collapse=" + ")
responder_formula <- paste("meth ~ responder +", cpPC_string)
```

```{r responder-ewas-TEST, eval=F}
run_test_model <- function(probeData, covarData, formula_string) {
  CpG <- rownames(probeData)
  regData <- cbind(meth=as.vector(probeData), covarData)

  outlier_TF <- (regData$meth < quantile(regData$meth, 0.25) - 3 * IQR(regData$meth) | 
                   (regData$meth > quantile(regData$meth, 0.75) + 3 * IQR(regData$meth)))
  regData <- regData[!outlier_TF, ]
  
  # regData$meth <- lm(meth ~ as.factor(Sentrix_ID), data=regData)$residuals
  # regData$meth <- lm(meth ~ cpPC2 + cpPC7 + cpPC9 + cpPC10, data=regData)$residual
    
  tryCatch({
    lm_fit <- lm(as.formula(formula_string), data=regData)
    # sandwich_fit <- coeftest(lm_fit, vcov=vcovHC(lm_fit, type="HC"))
    # coefs <- sandwich_fit["baseline_K", c("Estimate", "t value", "Pr(>|t|)")]
    coefs <- summary(lm_fit)$coef["response_K", c("Estimate", "t value", "Pr(>|t|)")]
    setNames(c(CpG, coefs), c("CpG", "beta", "t", "p"))
    
    # regData$Sentrix_ID <- as.factor(regData$Sentrix_ID)
    # lmm_fit <- lmer(meth ~ baseline_K + (1|Sentrix_ID), data=regData)
    # coefs <- summary(lmm_fit)$coef["baseline_K", c("Estimate", "t value", "Pr(>|t|)")]
    setNames(c(CpG, coefs), c("CpG", "beta", "t", "p"))
  }, error=function(e) rep(NA, 3))
}


cpPC_string <- paste(paste0("PC", c(1:5)), collapse=" + ")
responder_formula <- paste("meth ~ responder + cpPC9 + ", cpPC_string)
# responder_formula <- "meth ~ responder + cpPC2 + cpPC7 + cpPC9 + cpPC10"
# responder_formula <- "meth ~ responder + (Sentrix_ID == 202835220031) + cpPC2 + cpPC7 + cpPC9 + cpPC10"# + (Sentrix_ID == 202905580100) + (Sentrix_ID == 202905570138)"
# responder_formula <- "meth ~ baseline_K"
# responder_formula <- "meth ~ responder + cpPC2 + cpPC8 + cpPC9 + cpPC10"
responder_formula <- "meth ~ response_K + sex + CD4T + CD8T + Mono + Bcell + NK"
# small_sample <- sample(1:nrow(mvals), size=4000)

# mysubset <- nmd$Sentrix_ID != 202835220031 .... 202905580100 .... 202905570138
mysubset <- 1:nrow(nmd)
test_res_list <- foreach(
  meth=iter(mvals[small_sample, mysubset], by="row")) %do%
  run_test_model(meth, nmd[mysubset, ], responder_formula)

test_res_df <- do.call(rbind, test_res_list) %>%
  data.frame(stringsAsFactors=F) %>%
  mutate_at(c("beta", "t", "p"), as.numeric) %>%
  inner_join(dplyr::select(epic_anno, Name, chr, pos, UCSC_RefGene_Name), 
             by=c("CpG"="Name")) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>% 
  arrange(p)
gControl(test_res_df$p)

make_qqplot(test_res_df$p, plotTitle=gControl(test_res_df$p))

# ggplot(nmd, aes(x=responder, y=mvals["cg24002183", ], color=Mono)) + geom_point()
lm_fit <- lm(testm ~ responder + Gran, data=nmd)
summary(lm_fit)$coef
plot(lm_fit)

broom::tidy(lm(testm ~ cpPC10 + responder, data=nmd))

qplot(x=nmd$PC1, y=nmd$PC2, color=as.factor(nmd$Sentrix_ID))

hist(test_res_df$p)

ggplot(nmd, aes(color=as.factor(Sentrix_ID), x=cpPC10, y=cpPC9)) + geom_point(size=10)

ggplot(nmd, aes(y=lm(mvals["cg06618957",] ~ as.factor(Sentrix_ID), data=regData)$residuals, x=as.factor(Sentrix_ID))) +
  geom_point()

lmp <- function (modelobject) {
    if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
    f <- summary(modelobject)$fstatistic
    p <- pf(f[1],f[2],f[3],lower.tail=F)
    attributes(p) <- NULL
    return(p)
}

for (i in 1:10) {
  lm1 <- lm(as.formula(paste0("cpPC", i, " ~ as.factor(Sentrix_ID)")), data=nmd)
  print(broom::tidy(lm1))
  print(lmp(lm1))
}

for (i in 1:20) {
  lm1 <- lm(as.formula(paste0("cpPC", i, " ~ responder")), data=nmd)
  print(lmp(lm1))
}
```

```{r responder-ewas, cache=1}
# cl <- makePSOCKcluster(32)
# registerDoParallel(cl)
# 
# responder_ewas_res_list <- foreach(
#   meth=iter(mvals, by="row")) %dopar%
#   run_responder_model(meth, nmd, responder_formula)
# 
# stopCluster(cl)
lazyLoad("../cache/responder-ewas_bcb5c2ceea489b9379a74770a4bb1b23")
```

```{r responder-ewas-processing}
responder_res_df <- do.call(rbind, responder_ewas_res_list) %>%
  data.frame(stringsAsFactors=F) %>%
  mutate_at(c("beta", "t", "p"), as.numeric) %>%
  inner_join(dplyr::select(epic_anno, Name, chr, pos, UCSC_RefGene_Name), 
             by=c("CpG"="Name")) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>% 
  arrange(p)
```

```{r responder-ewas-adjusted, cache=1}
# cl <- makePSOCKcluster(32)
# registerDoParallel(cl)
# 
# responder_adj_formula <- paste0(responder_formula, " + age + sex")
# responder_adj_ewas_res_list <- foreach(
#   meth=iter(mvals, by="row")) %dopar%
#   run_responder_model(meth, nmd, responder_adj_formula)
# 
# stopCluster(cl)
lazyLoad("../cache/responder-ewas-adjusted_6c7157fca000c9120bdf578c58e82264")
```

```{r responder-ewas-adjusted-processing}
responder_adj_res_df <- do.call(rbind, responder_adj_ewas_res_list) %>%
  data.frame(stringsAsFactors=F) %>%
  mutate_at(c("beta", "t", "p"), as.numeric) %>%
  inner_join(dplyr::select(epic_anno, Name, chr, pos, UCSC_RefGene_Name), 
             by=c("CpG"="Name")) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>% 
  arrange(p)

 manhattan(filter(responder_adj_res_df, p<0.05), chr="chr", bp="pos", p="p", main="Age/sex adjust")
```

```{r responder-ewas-adjusted-cc, cache=1}
cl <- makePSOCKcluster(32)
registerDoParallel(cl)

main_p_nominal_cpgs <- responder_res_df$CpG[responder_res_df$p < 0.01]
responder_adjCC_formula <- paste0(responder_formula, " + CD8T + CD4T + Bcell + NK + Mono + Gran")
responder_adjCC_ewas_res_list <- foreach(
  meth=iter(mvals[main_p_nominal_cpgs, ], by="row")) %dopar%
  run_responder_model(meth, nmd, responder_adjCC_formula)

stopCluster(cl)
```

```{r responder-ewas-adjusted-cc-processing}
responder_adjCC_res_df <- do.call(rbind, responder_adjCC_ewas_res_list) %>%
  data.frame(stringsAsFactors=F) %>%
  mutate_at(c("beta", "t", "p"), as.numeric) %>%
  inner_join(dplyr::select(epic_anno, Name, chr, pos, UCSC_RefGene_Name),
             by=c("CpG"="Name")) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)

 manhattan(responder_adjCC_res_df, chr="chr", bp="pos", p="p", main="Cell count adjust")
```

```{r responder-ewas-adjusted-biol, cache=1}
cl <- makePSOCKcluster(32)
registerDoParallel(cl)

main_p_nominal_cpgs <- responder_res_df$CpG[responder_res_df$p < 0.01]
responder_adjBiol_formula <- paste0(responder_formula, " + age + sex + bmi_0 + TRIG")
responder_adjBiol_ewas_res_list <- foreach(
  meth=iter(mvals[main_p_nominal_cpgs, ], by="row")) %dopar%
  run_responder_model(meth, nmd, responder_adjBiol_formula)

stopCluster(cl)
```

```{r responder-ewas-adjusted-biol-processing}
responder_adjBiol_res_df <- do.call(rbind, responder_adjBiol_ewas_res_list) %>%
  data.frame(stringsAsFactors=F) %>%
  mutate_at(c("beta", "t", "p"), as.numeric) %>%
  inner_join(dplyr::select(epic_anno, Name, chr, pos, UCSC_RefGene_Name),
             by=c("CpG"="Name")) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)

 manhattan(responder_adjBiol_res_df, chr="chr", bp="pos", p="p", main="Age/sex/bmi/TG adjust")
```

```{r responder-ewas-model2, cache=1}
cl <- makePSOCKcluster(32)
registerDoParallel(cl)

main_p_nominal_cpgs <- responder_res_df$CpG[responder_res_df$p < 0.001]
responder_model2_formula <- paste0(responder_formula, " + age + sex + CD4T + CD8T + NK + Bcell + Mono")
responder_model2_ewas_res_list <- foreach(
  meth=iter(mvals[main_p_nominal_cpgs, ], by="row")) %dopar%
  run_responder_model(meth, nmd, responder_model2_formula)

stopCluster(cl)
```

```{r responder-ewas-adjusted-model2}
responder_model2_res_df <- do.call(rbind, responder_model2_ewas_res_list) %>%
  data.frame(stringsAsFactors=F) %>%
  mutate_at(c("beta", "t", "p"), as.numeric) %>%
  inner_join(dplyr::select(epic_anno, Name, chr, pos, UCSC_RefGene_Name),
             by=c("CpG"="Name")) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)

 manhattan(responder_model2_res_df, chr="chr", bp="pos", p="p", main="Age/sex/cell counts adjust")
```

```{r responder-ewas-model3, cache=1}
cl <- makePSOCKcluster(32)
registerDoParallel(cl)

main_p_nominal_cpgs <- responder_res_df$CpG[responder_res_df$p < 0.001]
responder_model3_formula <- paste0(responder_formula, " + age + sex + CD4T + CD8T + NK + Bcell + Mono + TRIG")
responder_model3_ewas_res_list <- foreach(
  meth=iter(mvals[main_p_nominal_cpgs, ], by="row")) %dopar%
  run_responder_model(meth, nmd, responder_model3_formula)

stopCluster(cl)
```

```{r responder-ewas-adjusted-model3}
responder_model3_res_df <- do.call(rbind, responder_model3_ewas_res_list) %>%
  data.frame(stringsAsFactors=F) %>%
  mutate_at(c("beta", "t", "p"), as.numeric) %>%
  inner_join(dplyr::select(epic_anno, Name, chr, pos, UCSC_RefGene_Name),
             by=c("CpG"="Name")) %>%
  mutate(fdr=p.adjust(p, method="BH")) %>%
  arrange(p)

 manhattan(responder_model3_res_df, chr="chr", bp="pos", p="p", main="Age/sex/cell counts/TG adjust")
```

## Model

* Exposure = log-transformed blood phylloquinone concentrations
    - Concentrations below the level of detection set to 0.01 before log-transforming
* Outcome = methylation beta-values (~ fraction of cells methylated)
* Covariates: 10 control-probe principal components (see QC/preliminaries report)

## Basic output

```{r responder-outputs}
manhattan(responder_res_df, chr="chr", bp="pos", p="p")

lambda_responder <- gControl(responder_res_df$p)
make_qqplot(responder_res_df$p, 
            plotTitle=paste0("lambda: ", round(lambda_responder, 2)))

top_cpgs <- responder_res_df$CpG[1:4]
bind_rows(list(model1=filter(responder_res_df, CpG %in% top_cpgs),
               model2=filter(responder_model2_res_df, CpG %in% top_cpgs),
               model3=filter(responder_model3_res_df, CpG %in% top_cpgs)),
          .id="model") %>% 
  mutate(Gene=gsub(";.*", "", UCSC_RefGene_Name),
         res_string=paste0(round(beta, 2), " (", 
                           format(p, scientific=T, digits=3), ")")) %>%
  select(model, CpG, Gene, res_string) %>%
  spread(key=model, value=res_string) %>%
  setNames(c("CpG", "Annotated Gene", "Model 1", "Model 2", "Model 3")) %>%
  kable(booktabs=T,
        caption="Top hits from vitamin K response EWAS") %>%
  footnote(general=c("Regression results presented as: beta (p-value)",
                     "Model 1: adjusted for 10 control-probe principal components",
                     "Model 2: additionally adjusted for age, sex, and estimated cell count fractions",
                     "Model 3: additionally adjusted for plasma triglycerides"),
           general_title="")
```

```{r responder-investigation, eval=F}
bonferroni_cpgs <- filter(responder_res_df, p < 0.05 /
                            nrow(responder_res_df))
suggestive_cpgs <- filter(responder_res_df, p < 1e-5)
fdr_cpgs <- filter(responder_res_df, fdr<0.2)

gsa_tbl <- missMethyl::gometh(fdr_cpgs$CpG, 
                              collection="GO", array.type="EPIC") %>%
  arrange(P.DE)
```

## Interpretation

```{r responder-gwas-comparison}
ewas_ranges <- GRanges(seqnames=responder_res_df$chr,
                       ranges=IRanges(start=responder_res_df$pos,
                                      end=responder_res_df$pos))

par(mfrow=c(1,2))
# Define interesting loci from EWAS
fdr_cpgs <- filter(responder_res_df, p < 0.0001)
fdr_cpg_ranges <- GRanges(seqnames=fdr_cpgs$chr, 
                          ranges=IRanges(start=fdr_cpgs$pos-100, 
                                         end=fdr_cpgs$pos+100))
# --> are those loci enriched for GWAS signal?
cpg_to_snps_overlap <- findOverlaps(gwas_ranges, fdr_cpg_ranges)
# make_large_qqplot(hassan_gwas_anno[queryHits(cpg_to_snps_overlap), "P-value"][[1]],
#                   "EWAS CpGs p<1e-4 \n--> GWAS SNPs +/- 100bp")

# Define interesting loci from GWAS
hassan_gwas_anno$fdr <- p.adjust(hassan_gwas_anno$`P-value`, method="BH")
fdr_snps <- filter(hassan_gwas_anno, `P-value` < 1e-4)
fdr_snp_ranges <- GRanges(seqnames=fdr_snps$chr, 
                          ranges=IRanges(start=fdr_snps$pos-100, 
                                         end=fdr_snps$pos+100))
# --> are those loci enriched for EWAS signal?
snp_to_cpgs_overlap <- findOverlaps(ewas_ranges, fdr_snp_ranges)
responder_gc_df <- mutate(responder_res_df, 
                p=pchisq(qchisq(p, df=1, lower.tail=F) / lambda_responder,
                         df=1, lower.tail=F))
make_large_qqplot(responder_gc_df[queryHits(snp_to_cpgs_overlap), "p"],
                  "EWAS CpGs near (+/- 100kb) suggestive GWAS SNPs")

responder_gc_df[queryHits(snp_to_cpgs_overlap), ] %>%
  filter(p < 1e-2) %>%
  mutate(p=format(p, scientific=T, digits=3),
         Gene=gsub(";.*", "", UCSC_RefGene_Name)) %>%
  select(CpG, Chr=chr, Pos=pos, Gene, P=p) %>%
  kable(booktabs=T,
        caption="Interesting CpGs near suggestive GWAS SNPs") %>%
  kable_styling(full_width=F)
```

```{r responder-hypothesis-genes}
lipid_related <- c("NPC1L1", "APOA4", "APOA5", "APOE", "SORT1", "LPL")
k_metab_related <- c("CYP4F2", "CYP2C9", "GGCX", "VKORC")
## use the above as regular expression inputs
## look at QQ plots of CpGs near these genes
short_list_genes <- c(lipid_related, k_metab_related)
short_list_cpgs <- filter(epic_anno, grepl(paste0("^", short_list_genes,
                                                  collapse="|"),
                                           UCSC_RefGene_Name))$Name
short_list_res <- filter(responder_gc_df, CpG %in% short_list_cpgs)

long_list_cpgs <- read_csv("../data/VK_genes_to_CpG.csv")$Name
long_list_res <- filter(responder_gc_df, CpG %in% long_list_cpgs)

par(mfrow=c(1,2))
make_large_qqplot(short_list_res$p, "Hypothesis CpGs (short list)")
make_large_qqplot(long_list_res$p, "Hypothesis CpGs (full list)")

responder_gc_df %>%
  filter(CpG %in% short_list_cpgs,
         -log10(p) > 1.5) %>%
  mutate(p=format(p, scientific=T, digits=3),
         Gene=gsub(";.*", "", UCSC_RefGene_Name)) %>%
  select(CpG, Chr=chr, Pos=pos, Gene, P=p) %>%
  kable(booktabs=T,
        caption="Interesting CpGs annotated to lipid- or vitamin K-related genes") %>%
  kable_styling(full_width=F)
```

<!-- ### Specific CpGs of interest -->

```{r responder-top-cpgs}
responder_res_df %>%
  head() %>%
  mutate(p=format(p, scientific=T, digits=3),
         Gene=gsub(";.*", "", UCSC_RefGene_Name)) %>%
  select(CpG, Chr=chr, Pos=pos, Gene, P=p) %>%
  kable(booktabs=T,
        caption="Top 20 CpGs from responder EWAS") %>%
  kable_styling(full_width=F)
```

Of note:

* Multiple genes associated with these top CpGs interact with or relate to receptors for vitamins A & D.
* `r responder_res_df$CpG[20]` is in CCDC12, which is known to interact with VKORC1. 
* 2 of the CpGs are near genes related to androgen metabolism.







